dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+  | otherwise  = modify (FreeVar n ==> m <>)
+```
+We use the occurence checker from [#Substitution](#Substitution) to check if the resulting type does not refer to itself. How to combine dependent and recursive types is pretty open question.
+
+A bound variable can only unify with itself. It represents some value or type that is not known yet - or can vary - in this region of the program.
+```haskell
+unify (Bound t) (Bound u) | t == u = mempty
+```
+
+For some structures unification is just a descent.
+```haskell
+unify (App a b) (App c d) = do
+  unify a c
+  unify b d
+
+unify Star Star = do
+  return ()
+
+unify (Lit l) (Lit k)
+  | l == k = mempty
+```
+
+For functions and their types though, it is a bit more complicated process. We need to unify the type of the argument and then replace the argument in the body of function (or dependent function type) with a name from its counterpart and only then we can unify it.
+```haskell
+unify (Pi n t b) (Pi m u c) = do
+  unify t u
+  unify b (subst (Bound m ==> Rigid n) c)
+
+unify (Lam n t b) (Lam m u c) = do
+  unify t u
+  unify b (subst (Bound m ==> Rigid n) c)
+```
+
+Another intesting thing is record and its type. For record, we unify given signatures of two records, and then their field values. For types of records we only unify types.
+```haskell
+unify (Record ds) (Record gs) = do
+  case zipDecls ds gs of
+    Just quads -> for_ quads \(tn, tm, n, m) -> do
+      unify tn tm  -- unify inferrent types of fields
+      unify  n  m  -- unify the fields themselves
+
+    Nothing -> do
+      die $ Mismatch (Record ds) (Record gs)
+
+unify (Product tds) (Product tgs) = do
+  case zipTDecls tds tgs of
+    Just pairs -> for_ pairs \(n, m) -> do
+      unify n m
+
+    Nothing -> do
+      die $ Mismatch (Product tds) (Product tgs)
+```
+
+Sadly, we can do nothing with _projections_. It is possible to have a type expression `a.b`, where `a` is a bound variable and `b` is a field name. It is not reasonable to unify `a ~ b` for `a.x ~ b.x`, because it is possible to have latter without the former. We will consider this an error. Technically, there are ways around it - by handling special cases - but that's not a general solution.
+```haskell
+unify a@Match {} b = die $ ProjectionL a b
+unify a b@Match {} = die $ ProjectionR a b
+
+unify a@Get {} b = die $ ProjectionL a b
+unify a a@Get {} = die $ ProjectionR a b
+```
+
+The `let`-expression of both kinds should not enter unifier at all. They can always be normalised to non-`let`-expressions.
+```haskell
+unify Let {} _ = error "unifier: left argument is a let-expression"
+unify _ Let {} = error "unifier: right argument is a let-expression"
+
+unify LetRec {} _ = error "unifier: left argument is a let-expression"
+unify _ LetRec {} = error "unifier: right argument is a let-expression"
+```
+
+And the last case is fired after all other fail, which means that types don't match.
+```haskell
+unify a b = die $ Mismatch a b
+```
+
+## Normalization
+
```

Comment:
You wrote 'Normalization', but you 'normalised' above. Choose your favourite variant for consistency.